src/scripts/main.js

```diff
@@ -1,3 +1,258 @@
+
+  while (!found) {
+    const r = Math.floor(Math.random() * rows);
+    const c = Math.floor(Math.random() * columns);
+
+    if (gameField[r][c] === 0) {
+      gameField[r][c] = randomizeNumber();
+      updateCell(field.rows[r].cells[c], gameField[r][c]);
+      found = true;
+    }
+  }
+}
+
+function hasEmptyCell() {
+  for (let r = 0; r < rows; r++) {
+    for (let c = 0; c < columns; c++) {
+      if (gameField[r][c] === 0) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+// 10% probability of the appearance of 4
+function randomizeNumber() {
+  return Math.random() >= 0.9 ? 4 : 2;
+}
+
+// updating cell value
+function updateCell(fieldCell, num) {
+  fieldCell.innerText = '';
+  fieldCell.className = 'field-cell';
+  fieldCell.classList.add(`field-cell--${num}`);
+
+  if (num > 0) {
+    fieldCell.innerText = num;
+  } else {
+    fieldCell.innerText = '';
+  }
+
+  if (!fieldCell.innerText) {
+    fieldCell.className = 'field-cell';
+  }
+
+  setGameOver();
+}
+
+// cells sliding
+function slide(row) {
+  let newRow = row;
+
+  newRow = filterZero(newRow);
+
+  for (let i = 0; i < newRow.length - 1; i++) {
+    if (newRow[i] === newRow[i + 1]) {
+      newRow[i] *= 2;
+      newRow[i + 1] = 0;
+      score += newRow[i];
+
+      if (newRow[i] === 2048) {
+        winMessage.classList.remove('hidden');
+      }
+    }
+  }
+
+  newRow = filterZero(newRow);
+
+  while (newRow.length < columns) {
+    newRow.push(0);
+  }
+
+  return newRow;
+}
+
+function filterZero(row) {
+  return row.filter(num => num !== 0);
+}
+
+function slideLeft() {
+  for (let r = 0; r < rows; r++) {
+    let row = gameField[r];
+
+    row = slide(row);
+    gameField[r] = row;
+
+    for (let c = 0; c < columns; c++) {
+      const num = gameField[r][c];
+
+      updateCell(field.rows[r].cells[c], num);
+    }
+  }
+}
+
+function slideRight() {
+  for (let r = 0; r < rows; r++) {
+    let row = gameField[r];
+
+    row.reverse();
+    row = slide(row);
+    row.reverse();
+    gameField[r] = row;
+
+    for (let c = 0; c < columns; c++) {
+      const num = gameField[r][c];
+
+      updateCell(field.rows[r].cells[c], num);
+    }
+  }
+}
+
+function slideUp() {
+  for (let c = 0; c < columns; c++) {
+    let row = [
+      gameField[0][c],
+      gameField[1][c],
+      gameField[2][c],
+      gameField[3][c],
+    ];
+
+    row = slide(row);
+
+    for (let r = 0; r < rows; r++) {
+      gameField[r][c] = row[r];
+
+      const num = gameField[r][c];
+
+      updateCell(field.rows[r].cells[c], num);
+    }
+  }
+}
+
+function slideDown() {
+  for (let c = 0; c < columns; c++) {
+    let row = [
+      gameField[0][c],
+      gameField[1][c],
+      gameField[2][c],
+      gameField[3][c],
+    ];
+
+    row.reverse();
+    row = slide(row);
+    row.reverse();
+
+    for (let r = 0; r < rows; r++) {
+      gameField[r][c] = row[r];
+
+      const num = gameField[r][c];
+
+      updateCell(field.rows[r].cells[c], num);
+    }
+  }
+}
+
```

Comment:
Use DRY principle. Think about how you can merge these functions into one.

Reply:
Thank you for your constructive review. Although it took me some time to think about and find the solution, it allowed me to make my code shorter and clearer!